lib/WasmReader/WasmByteCodeGenerator.cpp

```diff
@@ -1107,14 +1107,9 @@ WasmBytecodeGenerator::EmitReturnExpr(EmitInfo *lastStmtExprInfo)
         }
 
         m_writer.Conv(retOp, 0, retExprInfo.location);
-        if (!lastStmtExprInfo)
-        {
-            ReleaseLocation(&retExprInfo);
-        }
     }
     m_writer.AsmBr(m_funcInfo->GetExitLabel());
-
-    return EmitInfo();
+    return retExprInfo;
```

Comment:
I believe this isn't valid.
>The  `nop`, `if`, `br`, `br_if`, and  `return` constructs do not yield values
From: https://github.com/WebAssembly/design/blob/master/AstSemantics.md#yielding-values-from-control-constructs

It does prevent your example from working, but your example could easily be changed by moving one parenthesis,  like so:
    (func  (result i32) (if (i32.const 1) (return (i32.const 2))) (return (i32.const 3)))